browser/src/control/Control.UIManager.js

```diff
@@ -654,8 +655,19 @@ L.Control.UIManager = L.Control.extend({
 	showButton: function(buttonId, show) {
 		if (!this.notebookbar)
 			this.showButtonInClassicToolbar(buttonId, show);
-		else
+		else {
 			this.notebookbar.showShortcutsButton(buttonId, show);
+			if (show) {
+				delete this.hiddenButtons[buttonId];
+			} else {
+				this.hiddenButtons[buttonId] = true;
+			}
+			this.refreshNotebookbar();
```

Comment:
this operation has a big cost in sense of performance, it will reinitialize notebookbar in the core on the server, if we will add few buttons it will be done few times... (Control.Notebookbar.js `resetInCore`)

I wonder if we can avoid reset in the core and just rebuild HTML. It has to be tested with widgets like: style previews, font picker with previews - these notebookbar widgets are using core.

Reply:
I have taken that approach as:

1. being minimal in changes
2. meant to be a one time thing in the session
3. it worked

Reply:
I have reworked this. It should be faster.